group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +573,81 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     * @param offsetsRetentionMs The offset retention in milliseconds.
+     *
+     * @return The group id if the group no longer has any offsets remaining, empty otherwise.
+     */
+    public Optional<String> cleanupExpiredOffsets(String groupId, List<Record> records, long offsetsRetentionMs) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return Optional.of(groupId);
+        }
+        try {
+            Group group = groupMetadataManager.group(groupId);
+            ExpirationCondition expirationCondition = group.expirationCondition();
+            Set<TopicPartition> expiredPartitions = new HashSet<>();
+            long currentTimestamp = time.milliseconds();
+            AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+            offsetsByTopic.forEach((topic, partitions) -> {
+                if (!expirationCondition.subscribedTopics.contains(topic)) {
+                    partitions.forEach((partition, offsetAndMetadata) -> {
+                        OptionalLong expireTimestampMs = offsetAndMetadata.expireTimestampMs;
+                        if (expireTimestampMs.isPresent()) {
+                            // Older versions with explicit expire_timestamp field => old expiration semantics is used
+                            if (currentTimestamp >= expireTimestampMs.getAsLong()) {
+                                expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
+                            }
```

Comment:
Don't we need an else branch with `hasAllOffsetsExpired.set(false);` here? As a general advice, it would be great if we could wrap the logic to decide whether an offset is expired or not in a lambda/method. This would allow us to write something like the following here:

```
if (isOffsetExpired(...)) {
   expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
} else {
   hasAllOffsetsExpired.set(false);
}
```

This would handle all the cases. What do you think?

Reply:
Thanks for the catch and the suggestion

Reply:
thanks for the suggestion. looks much simpler